flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/dataformat/Timestamp.java

```diff
@@ -0,0 +1,202 @@
+	}
+
+	public int getNanoOfMillisecond() {
+		return nanoOfMillisecond;
+	}
+
+	/**
+	 * Obtains an instance of {@code Timestamp} from a millisecond.
+	 *
+	 * <p>This returns a {@code Timestamp} with the specified millisecond.
+	 * The nanoOfMillisecond field will be set to zero.
+	 *
+	 * @param millisecond the number of milliseconds since January 1, 1970, 00:00:00 GMT
+	 *                    A negative number is the number of milliseconds before
+	 *                    January 1, 1970, 00:00:00 GMT
+	 * @param precision the precision of fractional seconds
+	 * @return an instance of {@code Timestamp}
+	 */
+	public static Timestamp fromLong(long millisecond, int precision) {
+		long milli = isCompact(precision) ?
+			zeroLastNDigits(millisecond, 3 - precision) : millisecond;
+		return new Timestamp(milli, 0);
+	}
+
+	/**
+	 * Convert this {@code Timestmap} object to a {@link java.sql.Timestamp}.
+	 *
+	 * @return An instance of {@link java.sql.Timestamp}
+	 */
+	public java.sql.Timestamp toTimestamp() {
+		return java.sql.Timestamp.valueOf(toLocalDateTime());
+	}
+
+	/**
+	 * Obtains an instance of {@code Timestamp} from an instance of {@link java.sql.Timestamp}.
+	 *
+	 * <p>This returns a {@code Timestamp} with the specified {@link java.sql.Timestamp}.
+	 *
+	 * @param ts an instance of {@link java.sql.Timestamp}
+	 * @param precision the precision of fractional seconds
+	 * @return an instance of {@code Timestamp}
+	 */
+	public static Timestamp fromTimestamp(java.sql.Timestamp ts, int precision) {
+		return fromLocalDateTime(ts.toLocalDateTime(), precision);
+	}
+
+	/**
+	 * Convert this {@code Timestamp} object to a {@link LocalDateTime}.
+	 *
+	 * @return An instance of {@link LocalDateTime}
+	 */
+	public LocalDateTime toLocalDateTime() {
+		int date = (int) (millisecond / MILLIS_PER_DAY);
+		int time = (int) (millisecond % MILLIS_PER_DAY);
+		if (time < 0) {
+			--date;
+			time += MILLIS_PER_DAY;
+		}
+
+		int hour = time / (int) MILLIS_PER_HOUR;
+		int time2 = time % (int) MILLIS_PER_HOUR;
+		int minute = time2 / (int) MILLIS_PER_MINUTE;
+		int time3 = time2 % (int) MILLIS_PER_MINUTE;
+		int second = time3 / (int) MILLIS_PER_SECOND;
+		int nano = time3 % (int) MILLIS_PER_SECOND * 1_000_000;
+
+		long nanoOfDay = hour * NANOS_PER_HOUR +
+			minute * NANOS_PER_MINUTE +
+			second * NANOS_PER_SECOND +
+			nano +
+			nanoOfMillisecond;
+
+		LocalDate localDate = LocalDate.ofEpochDay(date);
+		LocalTime localTime = LocalTime.ofNanoOfDay(nanoOfDay);
+		return LocalDateTime.of(localDate, localTime);
+	}
+
+	/**
+	 * Obtains an instance of {@code Timestamp} from an instance of {@link LocalDateTime}.
+	 *
+	 * <p>This returns a {@code Timestamp} with the specified {@link LocalDateTime}.
+	 *
+	 * @param dateTime an instance of {@link LocalDateTime}
+	 * @param precision the precision of fractional seconds
+	 * @return an instance of {@code Timestamp}
+	 */
+	public static Timestamp fromLocalDateTime(LocalDateTime dateTime, int precision) {
```

Comment:
Why need `precision`?

Reply:
My original thought is we should zero last digits according to the precision. 
After discussing offline, we could hold this and remove `precision`.